source/common/buffer/buffer_impl.cc

```diff
@@ -148,21 +369,80 @@ Api::SysCallIntResult OwnedImpl::read(int fd, uint64_t max_length) {
 }
 
 uint64_t OwnedImpl::reserve(uint64_t length, RawSlice* iovecs, uint64_t num_iovecs) {
-  uint64_t ret = evbuffer_reserve_space(buffer_.get(), length,
-                                        reinterpret_cast<evbuffer_iovec*>(iovecs), num_iovecs);
-  ASSERT(ret >= 1);
-  return ret;
+  // TODO support the use of space in more than one slice.
+  if (num_iovecs == 0) {
+    return 0;
+  }
+  if (slices_.empty() || slices_.back()->reservableSize() < length) {
+    slices_.emplace_back(std::make_unique<OwnedBufferSlice>(length));
+  }
+  ASSERT(slices_.back()->reservableSize() >= length);
+  BufferSlice::Reservation reservation = slices_.back()->reserve(length);
+  iovecs[0].mem_ = reservation.first;
+  iovecs[0].len_ = reservation.second;
+  return 1;
 }
 
 ssize_t OwnedImpl::search(const void* data, uint64_t size, size_t start) const {
-  evbuffer_ptr start_ptr;
-  if (-1 == evbuffer_ptr_set(buffer_.get(), &start_ptr, start, EVBUFFER_PTR_SET)) {
-    return -1;
+  // This implementation uses the same search algorithm as evbuffer_search(), a naive
```

Comment:
I haven't seen this method show up in CPU profiles from production systems, so as far as I can tell it doesn't get used much in Envoy.

There are three calls to the method in Envoy currently, one of which appears to search untrusted inputs - and that one has a fixed pattern of length 12.

It's probably worth noting that, in the common case, both the current `evbuffer_search` implementation and this new code already benefit from an existing optimization: they use `memchr` to find the first character of the pattern, and for a large input size _n_, a typical vectorized `memchr` implementation only needs to do _n_/_v_ conditional branches, where _v_ is 16 or 32 depending on the size of the available SIMD registers.

Reply:
I was in error thinking this is easy to linearize, and I see the value in being able to do the vectorized memchr, so this looks like a good framework to start with until we see a perf problem.

But since the algorithm is very long with lots of corner-cases I think you need to add unit tests to cover a bunch of them. The current buffer_test.cc doesn't try to cover the underlying evbuffer_search impl which is assumed to be tested in the libevent repo, but now that you own it, you need to test it.